libsnark/zk_proof_systems/plonk/verifier.tcc

```diff
@@ -0,0 +1,620 @@
+/// at
+///   the secret input
+/// \param[out] S_polys_at_secret_g1: permutation polynomials S evaluated at the
+///   secret input
+template<typename ppT>
+verifier_preprocessed_input_t<ppT> plonk_verifier<ppT>::preprocessed_input(
+    const srs<ppT> &srs)
+{
+    verifier_preprocessed_input_t<ppT> preprocessed_input;
+    preprocessed_input.Q_polys_at_secret_g1.resize(srs.Q_polys.size());
+    plonk_evaluate_polys_at_secret_G1<ppT>(
+        srs.secret_powers_g1,
+        srs.Q_polys,
+        preprocessed_input.Q_polys_at_secret_g1);
+
+    preprocessed_input.S_polys_at_secret_g1.resize(srs.S_polys.size());
+    plonk_evaluate_polys_at_secret_G1<ppT>(
+        srs.secret_powers_g1,
+        srs.S_polys,
+        preprocessed_input.S_polys_at_secret_g1);
+    return preprocessed_input;
+}
+
+/// Verifier Step 1: validate that elements belong to group G1
+///
+/// \attention This validation MUST be done by the caller. Empty
+/// function here for consistency with the description in [GWC19]
+template<typename ppT>
+void plonk_verifier<ppT>::step_one(const plonk_proof<ppT> &proof)
+{
+}
+
+/// Verifier Step 2: validate that elements belong to scalar field Fr
+///
+/// \attention This validation MUST be done by the caller. Empty
+/// function here for consistency with the description in [GWC19]
+template<typename ppT>
+void plonk_verifier<ppT>::step_two(const plonk_proof<ppT> &proof)
+{
+}
+
+/// Verifier Step 3: validate that the public input belongs to scalar
+/// field Fr
+///
+/// \attention This validation MUST be done by the caller. Empty
+/// function here for consistency with the description in [GWC19]
+template<typename ppT> void plonk_verifier<ppT>::step_three(const srs<ppT> &srs)
+{
+}
+
+/// Verifier Step 4: compute challenges hashed transcript as in prover
+/// description, from the common inputs, public input, and elements of
+/// pi-SNARK. TODO: fixed to the test vectors for now
+///
+/// INPUT
+/// \param[in] transcript_hash: hashes of the communication transcript
+///            after prover rounds 1,2,3,4,5. TODO: \attention
+///            currently the hashes are pre-computed by the caller and
+///            passed as input for the purpouses of unit testing. In
+///            the long run this input can be removed and the hashes
+///            can be computed directly inside verifier::step_four()
+///
+/// OUTPUT
+/// \param[out] beta, gamma: permutation challenges - hashes of
+///             transcript
+/// \param[out] alpha: quotinet challenge - hash of transcript
+/// \param[out] zeta: evaluation challenge - hash of transcript
+/// \param[out] nu: opening challenge - hash of transcript (denoted by
+///             v in [GWC19])
+/// \param[out] u: multipoint evaluation challenge - hash of
+///             transcript
+template<typename ppT>
+step_four_out_t<ppT> plonk_verifier<ppT>::step_four(
+    transcript_hash_t<ppT> &transcript_hash)
+{
+    // step 4 output
+    libff::Fr<ppT> beta;
```

Comment:
This function declares, then copies, then moves.  I think you can just pass `Field(transcript_hash.beta)` directly to the constructor.

Reply:
Move semantics don't do anything for field elements, so `step_four_t` should just take const references, and you can just pass `transcript_hash.beta, ...` directly.

Reply:
Good points. Fixed.